File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/configuration/sdk_environment.dart
import 'package:runanywhere/core/native/rac_native.dart' show RacNative;
import 'package:runanywhere/foundation/errors/sdk_exception.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/generated/model_types.pbenum.dart'
    show SDKEnvironment;
import 'package:runanywhere/native/dart_bridge_environment.dart';

export 'package:runanywhere/generated/model_types.pbenum.dart'
    show SDKEnvironment;

extension SDKEnvironmentExtension on SDKEnvironment {
  /// C `rac_environment_t` value (RAC_ENV_DEVELOPMENT/STAGING/PRODUCTION).
  /// Mirrors Swift `cEnvironment` (SDKEnvironment.swift:52-59): unknown
  /// values map to RAC_ENV_DEVELOPMENT.
  int get _cEnvironment {
    switch (this) {
      case SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT:
        return 0; // RAC_ENV_DEVELOPMENT
      case SDKEnvironment.SDK_ENVIRONMENT_STAGING:
        return 1; // RAC_ENV_STAGING
      case SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION:
        return 2; // RAC_ENV_PRODUCTION
      default:
        return 0; // RAC_ENV_DEVELOPMENT
    }
  }

  /// Invoke a required `rac_env_*` predicate from commons.
  bool _envPredicate(bool Function(int)? fn, {required String symbol}) {
    if (fn == null) {
      throw UnsupportedError('$symbol is unavailable');
    }
    return fn(_cEnvironment);
  }

  String get description {
    switch (this) {
      case SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT:
        return 'Development Environment';
      case SDKEnvironment.SDK_ENVIRONMENT_STAGING:
        return 'Staging Environment';
      case SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION:
        return 'Production Environment';
      default:
        return 'Unspecified Environment';
    }
  }

  /// Check if this is a production environment (uses C++). Mirrors Swift
  /// `isProduction` (SDKEnvironment.swift:73).
  bool get isProduction => _envPredicate(
    RacNative.bindings.rac_env_is_production,
    symbol: 'rac_env_is_production',
  );

  /// Check if this is a testing environment (uses C++).
  bool get isTesting => _envPredicate(
    RacNative.bindings.rac_env_is_testing,
    symbol: 'rac_env_is_testing',
  );

  /// Check if this environment requires a valid backend URL (uses C++).
  bool get requiresBackendURL => _envPredicate(
    RacNative.bindings.rac_env_requires_backend_url,
    symbol: 'rac_env_requires_backend_url',
  );

  bool get isCompatibleWithCurrentBuild {
    switch (this) {
      case SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT:
      case SDKEnvironment.SDK_ENVIRONMENT_STAGING:
        return true;
      case SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION:
        var isDebug = false;
        assert(() {
          isDebug = true;
          return true;
        }());
        return !isDebug;
      default:
        return false;
    }
  }

  static bool get isDebugBuild {
    var isDebug = false;
    assert(() {
      isDebug = true;
      return true;
    }());
    return isDebug;
  }

  LogLevel get defaultLogLevel {
    switch (this) {
      case SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT:
        return LogLevel.LOG_LEVEL_DEBUG;
      case SDKEnvironment.SDK_ENVIRONMENT_STAGING:
        return LogLevel.LOG_LEVEL_INFO;
      case SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION:
        return LogLevel.LOG_LEVEL_WARNING;
      default:
        return LogLevel.LOG_LEVEL_INFO;
    }
  }

  /// Should send telemetry data (production only) — uses C++. Mirrors Swift
  /// `shouldSendTelemetry` (SDKEnvironment.swift:121).
  bool get shouldSendTelemetry => _envPredicate(
    RacNative.bindings.rac_env_should_send_telemetry,
    symbol: 'rac_env_should_send_telemetry',
  );

  /// Should sync with backend (non-development) — uses C++.
  bool get shouldSyncWithBackend => _envPredicate(
    RacNative.bindings.rac_env_should_sync_with_backend,
    symbol: 'rac_env_should_sync_with_backend',
  );

  /// Requires API authentication (non-development) — uses C++.
  bool get requiresAuthentication => _envPredicate(
    RacNative.bindings.rac_env_requires_auth,
    symbol: 'rac_env_requires_auth',
  );
}

class SupabaseConfig {
  final Uri projectURL;
  final String anonKey;

  SupabaseConfig({required this.projectURL, required this.anonKey});

  static SupabaseConfig? configuration(SDKEnvironment environment) {
    switch (environment) {
      case SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT:
        final supabaseUrl = DartBridgeDevConfig.supabaseURL;
        final supabaseKey = DartBridgeDevConfig.supabaseKey;

        // Delegate the placeholder + URL-shape checks to the canonical commons
        // rule (via DartBridge) so every SDK agrees instead of each carrying its
        // own regex.
        if (supabaseUrl == null ||
            supabaseKey == null ||
            !DartBridgeDevConfig.isUsableHttpUrl(supabaseUrl) ||
            !DartBridgeDevConfig.isUsableCredential(supabaseKey)) {
          return null;
        }

        final uri = Uri.tryParse(supabaseUrl);
        if (uri == null) {
          return null;
        }

        return SupabaseConfig(projectURL: uri, anonKey: supabaseKey);
      default:
        return null;
    }
  }
}

class SDKInitParams {
  final String apiKey;
  final Uri baseURL;
  final SDKEnvironment environment;

  SupabaseConfig? get supabaseConfig =>
      SupabaseConfig.configuration(environment);

  SDKInitParams({
    required this.apiKey,
    required this.baseURL,
    this.environment = SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION,
  }) {
    // Fail-closed on UNSPECIFIED / UNRECOGNIZED: refuse to silently coerce
    // an unknown wire-decoded enum value into DEVELOPMENT (which would
    // disable auth and backend sync).
    if (environment == SDKEnvironment.SDK_ENVIRONMENT_UNSPECIFIED ||
        !SDKEnvironment.values.contains(environment)) {
      throw SDKException.invalidConfiguration(
        'Unknown SDKEnvironment value; refusing to default to DEVELOPMENT',
      );
    }
    final result = DartBridgeEnvironment.instance.validateConfig(
      environment: environment,
      apiKey: apiKey,
      baseURL: baseURL.toString(),
    );
    if (!result.isValid) {
      throw SDKException.validationFailed(
        DartBridgeEnvironment.instance.getValidationErrorMessage(result),
        fieldPath: 'SDKInitParams',
      );
    }
  }

  factory SDKInitParams.fromString({
    required String apiKey,
    required String baseURL,
    SDKEnvironment environment = SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION,
  }) {
    final uri = Uri.tryParse(baseURL);
    if (uri == null) {
      throw ArgumentError('Invalid base URL: $baseURL');
    }
    return SDKInitParams(
      apiKey: apiKey,
      baseURL: uri,
      environment: environment,
    );
  }

  factory SDKInitParams.forDevelopment({String apiKey = ''}) {
    final supabaseConfig = SupabaseConfig.configuration(
      SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT,
    );
    return SDKInitParams(
      apiKey: apiKey,
      baseURL:
          supabaseConfig?.projectURL ??
          Uri.parse('https://dev.runanywhere.local'),
      environment: SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT,
    );
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Configuration/SDKEnvironment.swift
//
//  SDKEnvironment.swift
//  RunAnywhere SDK
//
//  SDK Environment mode — determines how data is handled.
//
//  `SDKEnvironment` is a typealias for the proto3-generated
//  `RASDKEnvironment` (idl/model_types.proto). The hand-written enum was
//  removed; all C-bridge helpers, validation, and UX behaviour are
//  preserved as extensions.
//

import CRACommons
import Foundation

// MARK: - Typealias

/// SDK Environment mode — determines how data is handled.
public typealias SDKEnvironment = RASDKEnvironment

// MARK: - Codable (wire format = lowercase)
//
// `wireString` and `from(wireString:)` are codegen-generated in
// Generated/RAConvenience.swift from the `rac_wire_string` annotations in
// idl/model_types.proto. The Codable conformance below delegates to those
// generated accessors so encoding and decoding share one canonical mapping.

extension RASDKEnvironment: Codable {
    public init(from decoder: Swift.Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = RASDKEnvironment.from(wireString: raw) ?? .unspecified
    }

    public func encode(to encoder: Swift.Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(self.wireString)
    }
}

// MARK: - Extensions (preserved from the hand-written enum)

public extension RASDKEnvironment {
    /// All three deployable environments, excluding `.unspecified` /
    /// `UNRECOGNIZED`. Preserves the `CaseIterable.allCases` semantics of
    /// the pre-IDL hand-written enum.
    static var deployableCases: [RASDKEnvironment] {
        [.development, .staging, .production]
    }

    // MARK: - C++ Bridge

    /// Convert to C++ environment type for cross-platform consistency.
    var cEnvironment: rac_environment_t {
        switch self {
        case .development: return RAC_ENV_DEVELOPMENT
        case .staging:     return RAC_ENV_STAGING
        case .production:  return RAC_ENV_PRODUCTION
        default:           return RAC_ENV_DEVELOPMENT
        }
    }

    /// Human-readable description.
    var description: String {
        switch self {
        case .development: return "Development Environment"
        case .staging:     return "Staging Environment"
        case .production:  return "Production Environment"
        default:           return "Unspecified Environment"
        }
    }

    /// Check if this is a production environment (uses C++).
    var isProduction: Bool { rac_env_is_production(cEnvironment) }

    /// Check if this is a testing environment (uses C++).
    var isTesting: Bool { rac_env_is_testing(cEnvironment) }

    /// Check if this environment requires a valid backend URL (uses C++).
    var requiresBackendURL: Bool { rac_env_requires_backend_url(cEnvironment) }

    // MARK: - Build Configuration Validation

    /// Check if the current build configuration is compatible with this
    /// environment. Production is only allowed in Release builds.
    var isCompatibleWithCurrentBuild: Bool {
        switch self {
        case .development, .staging:
            return true
        case .production:
            #if DEBUG
            return false
            #else
            return true
            #endif
        default:
            return false
        }
    }

    /// Returns true if we're running in a DEBUG build.
    static var isDebugBuild: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }

    // MARK: - Environment-Specific Settings

    /// Determine logging verbosity based on environment.
    var defaultLogLevel: RALogLevel {
        switch self {
        case .development: return .debug
        case .staging:     return .info
        case .production:  return .warning
        default:           return .info
        }
    }

    /// Should send telemetry data (production only) — uses C++.
    var shouldSendTelemetry: Bool { rac_env_should_send_telemetry(cEnvironment) }

    /// Should sync with backend (non-development) — uses C++.
    var shouldSyncWithBackend: Bool { rac_env_should_sync_with_backend(cEnvironment) }

    /// Requires API authentication (non-development) — uses C++.
    var requiresAuthentication: Bool { rac_env_requires_auth(cEnvironment) }
}

/// SDK initialization parameters.
public struct SDKInitParams: Sendable {
    /// API key for authentication.
    public let apiKey: String

    /// Base URL for API requests. Required for staging/production; optional
    /// for development (uses placeholder if not provided).
    public let baseURL: URL

    /// Environment mode (development/staging/production).
    public let environment: SDKEnvironment

    // MARK: - Default Development URL

    /// Placeholder URL used for development when no URL is provided.
    /// Development mode uses local analytics, so this is just a placeholder.
    private static let developmentPlaceholderURL: URL = {
        guard let url = URL(string: "https://dev.runanywhere.local") else {
            fatalError("Invalid hardcoded development URL")
        }
        return url
    }()

    // MARK: - Initializers

    /// Create initialization parameters for staging or production.
    public init(
        apiKey: String,
        baseURL: URL,
        environment: SDKEnvironment = .production
    ) throws {
        self.apiKey = apiKey
        self.baseURL = baseURL
        self.environment = environment

        try Self.validate(apiKey: apiKey, baseURL: baseURL, environment: environment)
    }

    /// Convenience initializer with string URL for staging or production.
    public init(
        apiKey: String,
        baseURL: String,
        environment: SDKEnvironment = .production
    ) throws {
        guard let url = URL(string: baseURL) else {
            throw SDKException(code: .validationFailed, message: "Invalid base URL format: \(baseURL)", category: .internal)
        }
        try self.init(apiKey: apiKey, baseURL: url, environment: environment)
    }

    /// Convenience initializer for development mode (no URL required).
    public init(forDevelopmentWithAPIKey apiKey: String = "") {
        self.apiKey = apiKey
        self.baseURL = Self.developmentPlaceholderURL
        self.environment = .development
    }

    // MARK: - Validation (Uses C++ for cross-platform consistency)

    private static func validate(
        apiKey: String,
        baseURL: URL,
        environment: SDKEnvironment
    ) throws {
        let logger = SDKLogger(category: "SDKInitParams")

        let cEnv = environment.cEnvironment

        let apiKeyResult = apiKey.withCString { ptr in
            rac_validate_api_key(ptr, cEnv)
        }
        if apiKeyResult != RAC_VALIDATION_OK {
            let message = String(cString: rac_validation_error_message(apiKeyResult))
            switch apiKeyResult {
            case RAC_VALIDATION_API_KEY_REQUIRED:
                throw SDKException(code: .invalidApiKey, message: "\(message) for \(environment.description)", category: .internal)
            case RAC_VALIDATION_API_KEY_TOO_SHORT:
                throw SDKException(code: .invalidApiKey, message: message, category: .internal)
            default:
                throw SDKException(code: .validationFailed, message: message, category: .internal)
            }
        }

        let urlResult = baseURL.absoluteString.withCString { ptr in
            rac_validate_base_url(ptr, cEnv)
        }
        if urlResult != RAC_VALIDATION_OK {
            let message = String(cString: rac_validation_error_message(urlResult))
            throw SDKException(code: .validationFailed, message: message, category: .internal)
        }

        if environment == .staging, baseURL.scheme?.lowercased() == "http" {
            logger.warning("Using HTTP for staging environment. Consider using HTTPS for security.")
        }

        if environment == .staging, let host = baseURL.host?.lowercased() {
            if host.contains("localhost") || host.contains("127.0.0.1") ||
               host.contains("example.com") || host.contains(".local") {
                logger.warning("Staging environment using local/example URL: \(host)")
            }
        }

        logger.info("URL validated for \(environment.description): \(baseURL.absoluteString)")
    }
}
